File: 03无重复字符的最长子串.cpp
#include <iostream>
#include <string>
#include <unordered_set>
using namespace std;

// 暴力解法
class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        int ret = 0;
        for (int i = 0; i < s.size(); i++)
        {
            unordered_set<char> se;
            int count = 0;
            for (int j = i; j < s.size(); j++)
            {
                if (!se.count(s[j]))
                {
                    se.insert(s[j]);
                    count++;
                }
                else
                    break;
            }
            ret = max(ret, count);
        }
        return ret;
    }
};

//  Sliding Window

class Solution2
{
public:
    int lengthOfLongestSubstring(string s)
    {
        if (s.size() == 0)
            return 0;
        int ret = 1, left = 0, right = 1;
        unordered_set<char> se;
        se.insert(s[0]);
        while (right < s.size())
        {
            while (se.count(s[right]))
            {
                se.erase(s[left]);
                left++;
            }
            se.insert(s[right]);
            ret = max(ret, right - left + 1);
            right++;
        }
        return ret;
    }
};

int main()
{
    Solution2 s;
    cout << s.lengthOfLongestSubstring("abaaaaaa") << endl;
}

File: 03无重复字符的最长子串.py
# 暴力解法
class Solution:

    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        ret = 0
        for i in range(n):
            count = 0
            se = set()
            for j in range(i, n):
                if s[j] in se:
                    break
                else:
                    se.add(s[j])
                    count += 1
            ret = max(ret, count)
        return ret


# 滑动窗口
class Solution2:

    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) == 0:
            return 0
        left = 0
        right = 1
        ret = 1
        se = {s[0]}
        while right < len(s):
            while s[right] in se:
                se.remove(s[left])
                left += 1
            se.add(s[right])
            ret = max(ret, right - left + 1)
            right += 1
        return ret


s = Solution2()
print(s.lengthOfLongestSubstring("abcabcac"))
